pydantic/_internal/_generate_schema.py

```diff
@@ -861,6 +836,9 @@ def _get_first_two_args_or_any(self, obj: Any) -> tuple[Any, Any]:
         return args[0], args[1]
 
     def _generate_schema_inner(self, obj: Any) -> core_schema.CoreSchema:
+        if _typing_extra.is_self(obj):
+            obj = self._resolve_self_type(obj)
+
```

Comment:
Just curious, why do we need to add this here now?

Reply:
Oops, seems like moving it breaks things, it needs to be right after the `__get_pydantic_core_schema__` check, so I'll leave it here